### PR TITLE
(PXP-10260) fix: reduce memory usage in database migration

### PR DESCRIPTION
**Pull request by paulineribeyre**
Jira Ticket: [PXP-10260](https://ctds-planx.atlassian.net/browse/PXP-10260)

### Improvements

* Reduce peak memory usage in database migrations

### CODE REVIEW

File: migrations/versions/4d93784a25e5_add_authz_column.py [hunk 1]
```diff
@@ -38,8 +38,10 @@ def upgrade():
 
     # extract existing PK (guid) and authz data (resource_path) from the metadata column
     connection = op.get_bind()
-    results = connection.execute("SELECT guid, data from metadata").fetchall()
-    for r in results:
+    query = "SELECT guid, data from metadata"
+    results = connection.execute(query)
+    r = results.fetchone()
```
Comment (paulineribeyre) [inline migrations/versions/4d93784a25e5_add_authz_column.py:43]: querying the records one by one might be a bit slow? we should be able to query a few at a time without impacting the memory usage negatively

File: migrations/versions/4d93784a25e5_add_authz_column.py [hunk 2]
```diff
@@ -38,8 +38,10 @@ def upgrade():
...
+    while r:
         guid, data = r[0], r[1]
         # default values for authz (["/open"])
         authz_data = json.loads(DEFAULT_AUTHZ_STR)
```
Comment (paulineribeyre) [inline migrations/versions/4d93784a25e5_add_authz_column.py:47]: let's take this line out of the loop

Comment (george42-ctds) [reply migrations/versions/4d93784a25e5_add_authz_column.py]: should the similar code in the downgrade also be removed? 
Like this `if` statement
`if authz_data.get(authz_key) != default_paths:`

File: migrations/versions/4d93784a25e5_add_authz_column.py [hunk 3]
```diff
@@ -38,28 +40,34 @@ def upgrade():
...
+    )
+    results = connection.execute(query).fetchall()
+    while results:
+        for r in results:
+            guid, data = r[0], r[1]
+            # check for any existing authz-resource-paths in the `data` column
+            if data is not None and authz_key in data:
+                authz_data[authz_key] = data.pop(authz_key)
+                # scrub internal fields from metadata
+                for metadata_key in remove_metadata_keys:
```
[the remark this answers is not in the document]
Comment (paulineribeyre) [reply migrations/versions/4d93784a25e5_add_authz_column.py:56]: we might be missing a unit test for this case?

[the remark this answers is not in the document]
Comment (paulineribeyre) [reply migrations/versions/4d93784a25e5_add_authz_column.py:56]: this is scope creep though :( if you agree about the use case i mentioned, i think we can file a bug ticket to address this separately

[the remark this answers is not in the document]
Comment (george42-ctds) [reply migrations/versions/4d93784a25e5_add_authz_column.py:56]: Oh, correct. Yes, I agree on all points. Could we make a new ticket for this?

[the remark this answers is not in the document]
Comment (paulineribeyre) [reply migrations/versions/4d93784a25e5_add_authz_column.py:56]: Sounds good. I can't approve my own PR, could you approve?

**Review by george42-ctds — APPROVED** [review]

File: migrations/versions/4d93784a25e5_add_authz_column.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (paulineribeyre) [reply migrations/versions/4d93784a25e5_add_authz_column.py:56]: Created ticket HP-908

**Pull request merged**